GVFS/FastFetch/NativeWindowsMethods.cs

```diff
@@ -0,0 +1,127 @@
+﻿using GVFS.Common.Tracing;
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Runtime.InteropServices;
+
+namespace FastFetch
+{
+    public class NativeWindowsMethods
+    {
+        private const int AccessDeniedWin32Error = 5;
+
+        public static unsafe void WriteFile(ITracer tracer, byte* originalData, long originalSize, string destination)
+        {
+            try
+            {
+                using (SafeFileHandle fileHandle = OpenForWrite(tracer, destination))
+                {
+                    if (fileHandle.IsInvalid)
+                    {
+                        throw new Win32Exception(Marshal.GetLastWin32Error());
+                    }
+
+                    byte* data = originalData;
+                    long size = originalSize;
+                    uint written = 0;
+                    while (size > 0)
+                    {
+                        uint toWrite = size < uint.MaxValue ? (uint)size : uint.MaxValue;
+                        if (!WriteFile(fileHandle, data, toWrite, out written, IntPtr.Zero))
+                        {
+                            throw new Win32Exception(Marshal.GetLastWin32Error());
+                        }
+
+                        size -= written;
+                        data = data + written;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                EventMetadata metadata = new EventMetadata();
+                metadata.Add("destination", destination);
+                metadata.Add("exception", e.ToString());
+                tracer.RelatedError(metadata, "Error writing file.");
+                throw;
+            }
+        }
+
+        public static bool StatAndUpdateIndexForFile(ITracer tracer, string path, MemoryMappedViewAccessor indexView, long offset)
+        {
+            try
+            {
+                FileInfo file = new FileInfo(path);
+                if (file.Exists)
+                {
+                    Index.IndexEntry indexEntry = new Index.IndexEntry(indexView, offset);
+                    indexEntry.Mtime = file.LastWriteTimeUtc;
+                    indexEntry.Ctime = file.CreationTimeUtc;
+                    indexEntry.Size = (uint)file.Length;
+                    return true;
+                }
+            }
+            catch (System.Security.SecurityException)
+            {
+                // Skip these.
+            }
+            catch (System.UnauthorizedAccessException)
+            {
+                // Skip these.
+            }
+
+            EventMetadata metadata = new EventMetadata();
+            metadata.Add("path", path);
+            tracer.RelatedError(metadata, "Error stat-ing file.");

```

Comment:
This means we'll log an error when the file does not exist, and I'm not sure that's what we want.

Based on how this method is called in `UpdateFileInformationForAllEntries` it appears that it's expected that it will sometimes return false.

I'm thinking the logging you added in `UpdateFileInformationFromDiskForFiles` should be sufficient, what do you think? (if you do decide to log something here, it should probably be Info rather than Error.

Reply:
I agree that the logging in `UpdateFileInformationFromDiskForFiles` should be sufficient here. I'll remove this.